testsuite/tests/singlecluster/gateway/reconciliation/listeners/test_gateway_basic_listeners.py

```diff
@@ -0,0 +1,52 @@
+"""
+Test case:
+- Add new listener and add it to HTTPRoute and test both work
+- Remove the new listener and remove it from HTTPRoute and test removed one is not working
+"""
+
+import pytest
+
+from testsuite.gateway import GatewayListenerTls
+from testsuite.utils import is_nxdomain, sleep_ttl
+
+
+pytestmark = [pytest.mark.kuadrant_only, pytest.mark.dnspolicy, pytest.mark.tlspolicy]
+
+
+@pytest.fixture(scope="module")
+def domain_prefix():
+    """Prefix for this test"""
+    return "listen"
+
+
+def test_listeners(custom_client, check_ok_https, gateway, route, wildcard_domain, second_domain):
+    """
+    This test checks reconciliation of dns/tls policy on addition and removal of listeners in gateway and HTTPRoute.
+    """
+
+    # Check the default domain works and second domain does not exist yet
+    check_ok_https(wildcard_domain)
+    assert is_nxdomain(second_domain)
+    assert custom_client(second_domain).get("/get").has_dns_error()
+
+    # Add second domain to gateway and route
+    gateway.refresh().add_listener(GatewayListenerTls(hostname=second_domain, gateway_name=gateway.name()))
+    gateway.apply()
+    route.refresh().add_hostname(second_domain)
+    route.apply()
+
+    # Check both domains work
+    for domain in [wildcard_domain, second_domain]:
+        check_ok_https(domain)
+
+    # Remove second domain
+    route.refresh().remove_hostname(second_domain)
+    route.apply()
+    gateway.refresh().remove_listener(second_domain)
+    gateway.apply()
+
+    # Check the default domain still works and second domain does not exist anymore
+    check_ok_https(wildcard_domain)
+    sleep_ttl(second_domain)
```

Comment:
Is this stable enough? If I got it right the `second_domain` should not exist at this point so there might be no wait at all. I don't really understand the internals of DNS discovery and propagation etc, but I think that it can happen that one DNS resolver does not resolve the domain anymore but some other resolver still might be able to resolve it because it takes some time for deletion to propagate.

When I tried this test I came across folowing error several times:
```
testsuite/tests/singlecluster/gateway/reconciliation/listeners/test_gateway_basic_listeners.py:52: in test_listeners
    assert custom_client(second_domain).get("/get").has_dns_error()
E   AssertionError: assert False
E    +  where False = has_dns_error()
```
This happend when running via `make testsuite/tests/singlecluster/gateway/reconciliation/listeners/test_gateway_basic_listeners.py`. If running via IDE (with or without debugging on) it always went ok.

Reply:
I encountered this problem before when running this test from Openshift pod. I am aware that in some circumstances this wait is not enough due to TTL caching of intermediate stub DNS resolvers.
Imagine scenario:
Original TTL of the record in authoritative DNS is 60
First caching DNS resolver queries the Authoritative DNS and stores the record and starts counting 60s
Second caching DNS resolver has a policy that caps TTL to 30 (This is a case on Openshift clusters) and queries First DNS server and starts counting 30
Finally testsuite queries for sleep_ttl() function the Second server and sleeps for 30s.

Now. After testsuite finishes sleeping, second DNS server has expired cache, authoritative DNS server does not contain the record as it was deleted, but first caching DNS server STILL HAS the record in cache! So testsuite now wants to check for dns error so it queries again, second server queries the first and gets an answer with ttl <30 but still valid. This results in the the failed assert. 

There are two ways I found out that can solve this problem. Use stub DNS resolvers that do not have policy capping the TTL or query the Authoritative DNS server for the origin TTL and use that value for sleep_ttl(). In our pipelines that run on Openshift we opted for first option. Implementing the second option can be difficult as some internal networks block port 53 so contacting authoritative DNS server is impossible from the testsuite.

**I will investigate this further.**

Reply:
Thanks for explanation! It is a pity that we just can't specify TTL in DNSPolicy. Nonetheless, would not it be sufficient to simply have 60s sleep hard-coded then? I assume we can reasonably expect that TTL is set to 60s so it would make the test stable and it is super simple solution. Might make more sense than investing time to lengthy investigation.

Reply:
Hard-coded sleep 60s is an easy option as from my observation this is the default TTL for DNS records created by DNSPolicy. But this can be changed in future. We can settle on this if there are no better solutions.

Reply:
I think I found better solution. We can actually get the authoritative TTL from DNSRecord CR in Openshift. But this must be fetched before the listener gets removed from the Gateway as that will delete the DNSRecord CR. What do you think?

Reply:
I added function `get_listener_dns_ttl()` to Gateway class, take a look.

Reply:
Nice!